pkgs/applications/window-managers/qtile/default.nix

```diff
@@ -17,34 +26,60 @@ python37Packages.buildPythonApplication rec {
   };
 
   patches = [
-    ./0001-Substitution-vars-for-absolute-paths.patch
+    (substituteAll {
+      src = ./0001-Substitution-vars-for-absolute-paths.patch;
+      glib = glib.out;
+      pango = pango.out;
+      inherit xcbutilcursor;
+      })
     ./0002-Restore-PATH-and-PYTHONPATH.patch
-    ./0003-Restart-executable.patch
+    (substituteAll {
+      src = ./0003-Restart-executable.patch;
+      out = "$out";
+      })
+    ./0004-Keep-env-in-test-process-spawner.patch #TODO upstream this
   ];
 
-  postPatch = ''
-    substituteInPlace libqtile/core/manager.py --subst-var-by out $out
-    substituteInPlace libqtile/pangocffi.py --subst-var-by glib ${glib.out}
-    substituteInPlace libqtile/pangocffi.py --subst-var-by pango ${pango.out}
-    substituteInPlace libqtile/backend/x11/xcursors.py --subst-var-by xcb-cursor ${xcbutilcursor.out}
-  '';
-
   SETUPTOOLS_SCM_PRETEND_VERSION = version;
 
   nativeBuildInputs = [ pkgconfig ];
-  buildInputs = [ glib libxcb cairo pango python37Packages.xcffib ];
+  buildInputs = [ glib libxcb cairo pango pypa.xcffib librsvg ]; 
+
+  pythonPath = with pypa; [ xcffib cairocffi-xcffib setuptools setuptools_scm ]; 
+
+  checkInputs = [ pypa.pytest pypa.pytest-rerunfailures xvfb_run xrandr xcalc xeyes xclock xterm pypa.psutil imagemagick ];
+
+  checkPhase = ''
+    runHook preCheck
 
-  pythonPath = with python37Packages; [ xcffib cairocffi-xcffib setuptools setuptools_scm ]; 
+    oldhome="$HOME"
+    HOME=$(mktemp -d) #tests need a home directory
+
+    # This is taken from scripts/ffibuild but excluding the pulse component, which we dont build,
+    echo "building pango"
+    python3 ./libqtile/pango_ffi_build.py
+    echo "building xcursors"
+    python3 ./libqtile/backend/x11/xcursors_ffi_build.py
+
+    patchShebangs /build/source/bin/*
+    export GDK_PIXBUF_MODULE_FILE="${librsvg.out}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache" #TODO is there a better way to do this?
+    #TODO figure out why tests fail nondeterministically, after fixing, pytest-rerunfailures will be unnecessary
+    # error: xcffib.ConnectionException: xcb connection errors because of socket, pipe and other stream errors.
+    xvfb-run -s '-screen 0 1024x768x24' pytest -vv --reruns 5
+
+    HOME="$oldhome"
+
+    runHook postCheck
+  '';
 
   postInstall = ''
     wrapProgram $out/bin/qtile \
-      --run 'export QTILE_WRAPPER=$0' \
-      --run 'export QTILE_SAVED_PYTHONPATH=$PYTHONPATH' \
-      --run 'export QTILE_SAVED_PATH=$PATH'
+      --set GDK_PIXBUF_MODULE_FILE "$GDK_PIXBUF_MODULE_FILE" \
+      --set QTILE_WRAPPER '$0' \
+      --set QTILE_SAVED_PYTHONPATH '$PYTHONPATH' \
+      --set QTILE_SAVED_PATH '$PATH'
```

Comment:
Haven't built it locally, but I think there's double wrapping here. If I'm correct, could you please use `makeWrapperArgs`?